libraries/standard/http/test/test-HTTPClient_Send.c

```diff
@@ -0,0 +1,228 @@
+                                     size_t bytesToRead )
+{
+    ( void ) pContext;
+    ( void ) pBuffer;
+
+    memcpy( pBuffer,
+            HTTP_TEST_RESPONSE_HEADER_LINES_NO_BODY,
+            sizeof( HTTP_TEST_RESPONSE_HEADER_LINES_NO_BODY ) - 1 );
+
+    return sizeof( HTTP_TEST_RESPONSE_HEADER_LINES_NO_BODY ) - 1;
+}
+
+/* Mocked network error returning transport send. */
+static int32_t transportSendNetworkError( HTTPNetworkContext_t * pContext,
+                                          const void * pBuffer,
+                                          size_t bytesToWrite )
+{
+    ( void ) pContext;
+    ( void ) pBuffer;
+    ( void ) bytesToWrite;
+
+    return -1;
+}
+
+/* Mocked transport send that returns less bytes than expected. */
+static int32_t transportSendLessThanBytesToWrite( HTTPNetworkContext_t * pContext,
+                                                  const void * pBuffer,
+                                                  size_t bytesToWrite )
+{
+    ( void ) pContext;
+    ( void ) pBuffer;
+    ( void ) bytesToWrite;
+
+    return bytesToWrite - 1;
+}
+
+
+int main()
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    HTTPResponse_t response = { 0 };
+
+    response.pBuffer = httpBuffer;
+    response.bufferLen = sizeof( httpBuffer );
+
+    HTTPTransportInterface_t transportInterface =
+    {
+        .recv     = transportRecvSuccess,
+        .send     = transportSendSuccess,
+        .pContext = NULL
+    };
+
+    HTTPRequestHeaders_t requestHeadersHead =
+    {
+        .pBuffer    = ( uint8_t * ) ( HTTP_TEST_REQUEST_HEAD ),
+        .bufferLen  = sizeof( HTTP_TEST_REQUEST_HEAD ) - 1,
+        .headersLen = sizeof( HTTP_TEST_REQUEST_HEAD ) - 1
+    };
+
+#define reset()                                                                \
+    do {                                                                       \
+        transportInterface.recv = transportRecvSuccess;                        \
+        transportInterface.send = transportSendSuccess;                        \
+        transportInterface.pContext = NULL;                                    \
+        requestHeadersHead.pBuffer = ( uint8_t * ) ( HTTP_TEST_REQUEST_HEAD ); \
+        requestHeadersHead.bufferLen = sizeof( HTTP_TEST_REQUEST_HEAD ) - 1;   \
+        requestHeadersHead.headersLen = sizeof( HTTP_TEST_REQUEST_HEAD ) - 1;  \
+    } while( 0 )
+    reset();
+
+    /* -----------------------------------------------------------------------*/
+
+    /* Happy Path testing. Test sending a request and successfully receiving a
+     * response that is within the bounds of the response buffer. */
+    returnStatus = HTTPClient_Send( &transportInterface,
+                                    &requestHeadersHead,
+                                    NULL,
+                                    0,
+                                    &response );
+
+    ok( returnStatus == HTTP_SUCCESS );
+    /* TODO: Check the response is parsed successfully. */
```

Comment:
Can we `DO` this `TODO` while we're working on this?  It's a problem approving new code with these, since they're not likely to appear in a future review unless the same code is affected.  We can add scanning for these, but in general it is best to just go ahead and implement at least part of these rather than leaving a `TODO`.

Reply:
It is an incremental review to make the review shorter. How about I just not leave a TODO word all?